DEV-675 - IAC Module concept / document (#2214)
gruntwork-io/docs: docs/2.0/docs/overview/concepts/iac-modules.md

# Leveraging Infrastructure as Code (IaC) Modules for Efficiency and Reliability
By understanding the different levels of abstraction in IaC and how modules fit into the broader picture, you can improve governance, documentation, MTTR (mean time to recovery), enhance consistency, and increase the reliability of your infrastructure.

## Understanding IaC Abstraction Levels
Gruntwork uses a four-layer system for describing infrastructure --

**Direct Infrastructure** (Lowest-Level): This is your actual deployed infrastructure, such as servers, networks, and databases.

> **Resource-Level IaC** (Mid-Level): OpenTofu provides wrappers around cloud provider APIs and represents objects in those APIs as [resources](https://opentofu.org/docs/language/resources/).  Using OpenTofu you can to define individual infrastructure resources (e.g., servers, networks, databases) in code. This improves consistency and enables version control, but managing complex deployments and establishing consistency and governance can still be challenging as everytime you deploy infrastructure you have to re-specify each resource and all its arguments.

Reviewer: ```suggestion
**Resource-Level IaC** (Mid-Level): OpenTofu provides wrappers around cloud provider APIs and represents objects in those APIs as [resources](https://opentofu.org/docs/language/resources/).  Using OpenTofu you can define individual infrastructure resources (e.g., servers, networks, databases) in code. This improves consistency and enables version control, but managing complex deployments and establishing consistency and governance can still be challenging given that every time you deploy infrastructure you have to re-specify each resource and all its arguments.
```

<img alt="Diagram showing IaC Module Hierarchy" className="img_node_modules-@docusaurus-theme-classic-lib-theme-MDXComponents-Img-styles-module medium-zoom-image" src="/img/iac/module_types.svg" />

## Benefits of IaC Modules
* **Increased Efficiency:** Modules eliminate the need to write repetitive code, accelerating infrastructure deployments.
* **Reduced Errors:** Pre-defined and tested modules minimize the risk of configuration errors and ensure consistency.
* **Improved Collaboration:** Modules promote code reuse and shared understanding across teams.
* **Enhanced Compliance:** Modules can incorporate security and compliance best practices, ensuring your infrastructure adheres to standards.
* **Simplified Documentation:** Modules serve as self-documenting building blocks, clearly defining infrastructure components and their relationships. Modules also regularly ship with separate documentation and examples to further explain, justify and demonstrate their use.

> * **Faster Recovery:** Modules stores in version control enable rapid redeployment of infrastructure from code, allowing you to quickly find and revert any mistakes from Click-ops, manual changes or other bad deployments.

Reviewer: ```suggestion
* **Faster Recovery:** Modules stored in version control enable rapid redeployment of infrastructure from code, allowing you to quickly find and revert any mistakes from Click-ops, manual changes or other bad deployments.
```

> * **Faster Scaling:** Modules allow you to scale infrastructure quickly by reusing existing building blocks to redeploy copies or in new regions/shared-fate-zones or data-centers.

Reviewer: ```suggestion
* **Faster Scaling:** Modules allow you to scale infrastructure quickly by reusing existing components to deploy additional copies or expand into new regions, shared-fate zones, or data centers.
```
